### ISSUE TITLE
Check for ambiguity in Bitcoin's serialization

### ISSUE THREAD

**Issue by defuse** [zcash/zcash#842]
In order to use an asymmetric proof of work (#27) we need to assume Bitcoin's serialization is unambiguous. That is, there's not a way to make two distinct sequences of bytes, both of which deserialize to the same block header. Otherwise we can use that ambiguity to grind the proof of work.

I don't think we *actually* have to make that assumption with the current code, since the PoW is checked by deserializing the block and then re-serializing it, and so there is no ambiguity. However we should spend some of our internal audit time checking this and other areas where there might be a way to grind the block header without changing the PoW input.

**Comment by ebfull** [comment]
I make sure that I add canonical representations, for example in my tree PR where I add serialization for `boost::optional`, to avoid this situation, under the assumption the Bitcoin folk would also do that. Hopefully they do.

In any case, for equihash the fields have no discriminants and so I'm not concerned about there being malleability attacks.

**Comment by nathan-at-least** [comment]
I added this to our roadmap and removed the "Engineering Meeting Agenda" label, assuming that's why the label was added. If not, @defuse should re-add the label and add a comment about what to discuss in the meeting.

**Comment by ebfull** [comment]
One area we need to explore is whether any of the fields in the block header use variable size integers and whether or not that can be used as an attack vector here.

**Comment by str4d** [comment]
The size field of the solution vector is a compact int, which can be one or three bytes. It's part of the unaltered serialisation of a vector from upstream.

**Comment by jl777** [comment]
https://bitcoin.org/en/developer-reference#compactsize-unsigned-integers

**Comment by daira** [comment]
@str4d and I looked at this in our pairing on Monday. (I had a long comment written about it but then had a power cut, so let's see what I can remember.)

The compactSize field just before the solution indeed *could* have been a problem, but we narrowly dodged a vulnerability. The issue is that in the case of a length that normally occupies 2 bytes, the 3- and 4-byte encodings are still available. The Bitcoin Developer Reference does not explicitly say that non-canonical encodings are rejected.

The input to the Equihash solver is generated based on the block header up to and including the nonce (as input to BLAKE2b). The difficulty check is done based on a SHA-256d hash of the whole block header. The remaining fields not included in the Equihash input are just the solution length and the solution. If we accepted non-canonical encodings for the solution length, then it would be possible to find an Equihash solution and then have three chances at passing the difficulty check: one for each possible encoding of the length. This would give miners using this attack a 3x advantage over miners not using it. (The same kind of attack is not possible on the solution itself, because the Equihash design precludes it by sorting solutions.)

However, this is not actually exploitable because the [ReadCompactSize](https://github.com/zcash/zcash/blob/master/src/serialize.h#L278) function (unchanged from upstream) does reject non-canonical encodings. We should probably have a test of this, and we need to document it in the protocol spec. I have opened https://github.com/zcash/zips/issues/70 for the latter. I will also open a documentation bug against Bitcoin since it's a consensus critical omission in their documentation.
